### ISSUE TITLE
[THREAD] PowerShell Generator - Quick Questions

### ISSUE THREAD

**Issue by fearthecowboy** [Azure/autorest.powershell#153]
If you have some quick questions about the PowerShell Generator, ask 'em here. 

If you have an actual bug to report, go ahead and file an issue.

**Comment by deepak2016** [comment]
@fearthecowboy has the updates been released? I am facing issue while compiling the generated module. 
```
PS D:\workspace\powershell\graph> .\generated\build-module.ps1
Creating isolated process...
Cleaning build folders...
Compiling module...
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphDriveAllof1.json.cs(56,217): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphDriveAllof1.json.cs(56,248): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphDriveAllof1.json.cs(59,225): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphDriveAllof1.json.cs(59,256): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphListAllof1.json.cs(54,232): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphListAllof1.json.cs(54,263): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphListAllof1.json.cs(55,247): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphListAllof1.json.cs(55,278): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphListAllof1.json.cs(58,216): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphListAllof1.json.cs(58,247): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
D:\workspace\powershell\graph\generated\build-module.ps1 : Compilation failed.
At D:\workspace\powershell\graph\generated\build-module.ps1:70 char:3
+   Write-Error 'Compilation failed.'
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
+ FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,build-module.ps1
```

**Comment by NakulSabharwal** [comment]
@fearthecowboy  I am also facing this issue while compiling. Any update on its fix ? 
```java
PS D:\workspace\powershell\graph> .\generated\build-module.ps1
Creating isolated process...
Cleaning build folders...
Compiling module...
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphDriveAllof1.json.cs(56,217): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphDriveAllof1.json.cs(56,248): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphDriveAllof1.json.cs(59,225): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphDriveAllof1.json.cs(59,256): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphListAllof1.json.cs(54,232): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphListAllof1.json.cs(54,263): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphListAllof1.json.cs(55,247): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphListAllof1.json.cs(55,278): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphListAllof1.json.cs(58,216): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
generated\api\Sample\API\Models\ComponentsSchemasMicrosoftGraphListAllof1.json.cs(58,247): error CS1061: 'IMicrosoftGraphSystemFacet' does not contain a definition for 'Linq' and no accessible extension method 'Linq' accepting a first argument of type 'IMicrosoftGraphSystemFacet' could be found (are you missing a using directive or an assembly reference?) [D:\workspace\powershell\graph\generated\Partial Graph API.csproj]
D:\workspace\powershell\graph\generated\build-module.ps1 : Compilation failed.
At D:\workspace\powershell\graph\generated\build-module.ps1:70 char:3
+   Write-Error 'Compilation failed.'
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
+ FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,build-module.ps1
```

**Comment by darrenjrobinson** [comment]
This may seem like a rudimentary question, but where should Directives be placed for generation of PowerShell Modules ?
I built a simple API and AutoRest successfully builds the PowerShell Module, but names it Invoke-Cmdletname when I'd like it to be Get-Cmdletname

Parameter Rename as detailed here https://github.com/Azure/autorest/blob/master/docs/powershell/options.md#Tweak-The-Way-It-Generates-Cmdlets has likely syntax however when using AutoRest to generate the Module the following error is thrown

ERROR: Schema violation: Additional properties not allowed: directive

Thinking I'm missing something quite basic and just need to be pointed in the right direction.

**Comment by fearthecowboy** [comment]
@darrenjrobinson  - Are you using a [configuration file](https://github.com/Azure/autorest/blob/master/docs/user/configuration.md)?

AutoRest allows extensive configuration thru it's configuration file, in which you stop putting stuff on the cmdline, and stick it in a file. 

You can use `.json`, `.yaml` or `.md` (markdown literate configuration) for your file. 

Simplest  make a yaml file : 

save as `configuration.yaml`

``` yaml
# your swagger file
input-file: foo.json

powershell: 
  namespace: MyCompany.API
  output-folder: out


directive:
  - where: 
      subject: VirtualMachine
      verb: Get
    set: 
      subject: VM         
```

Then from the command line: `autorest configuration.yaml`

---
We use the markdown literate configuration file format, which is a markdown file. 
This has the virtue of being both docs and configuration, and when you're navigating in github, the `readme.md` is rendered automatically, and autorest will look for `readme.md` in the current folder, so you don't even have to specify it on the cmdline

The only difference is that configuration is in YAML code blocks in the markdown file.

~~~ markdown
# Sample Configuration

_the following line is required!_
> see https://aka.ms/autorest

``` yaml
# your swagger file
input-file: foo.json

powershell: 
  namespace: MyCompany.API
  output-folder: out

directive:
  - where: 
      subject: VirtualMachine
      verb: Get
    set: 
      subject: VM 
```
~~~

**Comment by darrenjrobinson** [comment]
Hey @fearthecowboy yes I'm using a Config file.

Even using a module wide directive (at the top of the config file) such as the following fails. Also tried RegEx versions. 

directive:
  - where:
      verb: invoke
    set:
      verb: get

the output is 

Process() cancelled due to exception : path.match is not a function / TypeError: path.match is not a function
    at isAbsolute (/node_modules/@microsoft.azure/uri/dist/main.js:71:19)
    at Object.ResolveUri (/node_modules/@microsoft.azure/uri/dist/main.js:149:9)
    at ConfigurationView.ResolveAsPath (C:\Users\Darren Robinson\.autorest\@microsoft.azure_autorest-core@3.0.5402\node_modules\@microsoft.azure\autorest-core\dist\lib\configuration.js:236:22)
    at linq_es2015_1.From.Select.each (C:\Users\Darren Robinson\.autorest\@microsoft.azure_autorest-core@3.0.5402\node_modules\@microsoft.azure\autorest-core\dist\lib\configuration.js:305:34)
    at EnumerableImpl.Select [as _factory] (/node_modules/linq-es2015/lib/generators.js:29:15)
    at Select.next (<anonymous>)
    at EnumerableImpl.ToArray (/node_modules/linq-es2015/lib/linq.js:372:18)
    at ConfigurationView.get InputFileUris [as InputFileUris] (C:\Users\Darren Robinson\.autorest\@microsoft.azure_autorest-core@3.0.5402\node_modules\@microsoft.azure\autorest-core\dist\lib\configuration.js:306:14)
    at Object.get (C:\Users\Darren Robinson\.autorest\@microsoft.azure_autorest-core@3.0.5402\node_modules\@microsoft.azure\autorest-core\dist\lib\configuration.js:223:82)
    at processInternal (C:\Users\Darren Robinson\.autorest\@microsoft.azure_autorest-core@3.0.5402\node_modules\@microsoft.azure\autorest-core\dist\lib\autorest-core.js:103:26)
    at process._tickCallback (internal/process/next_tick.js:68:7)
  TypeError: path.match is not a function

**Comment by fearthecowboy** [comment]
Can you give us a concrete repro? I'll have to have @NelsonDaniel  take a look at it.

**Comment by darrenjrobinson** [comment]
@NelsonDaniel @fearthecowboy 

It a very simple API that I quickly built to help me get up to speed with AutoRest.
A very simple Get.
Here is the OpenAPI Definition that AutoRest successfully generates the module for but with the Verb Invoke, that I'm trying to change the verb to Get through the Config file.

swagger: '2.0'
info:
  title: BOFH
  version: 1.0.0
host: my.webhost.com.au
basePath: /
schemes:
  - https
  - http
paths:
  /api/excuse:
    get:
      operationId: /api/excuse/get
      produces:
        - application/json
      consumes: []
      parameters: []
      description: Get a BOFH Excuse
      responses:
        '200':
          description: Success operation
          schema:
            $ref: "#/definitions/GetExcuseResponseModel"
      security:
        - apikeyQuery: []
definitions:
  GetExcuseResponseModel:
    type: object
    properties:
      excuse:
        type: string
      excuseGenerated:
        type: string
securityDefinitions:
  apikeyQuery:
    type: apiKey
    name: code
    in: query

**Comment by JustinGrote** [comment]
Great work on this. I'm stepping through getting it to work for the LogicMonitor API since it has a swagger specification. One thing is that they have a convoluted authentication system requiring a complicated signature. I noticed that your example shows how to add custom functionality via C#, but sadly I'm a powershell guy :)

I did manage to use HttpPipelinePrepend to get what I want using your Gist example, but is there a better way to add this authorization to the commands?

```powershell
Get-Alertlist -HttpPipelinePrepend { param($req, $callback, $next ) 
    <# Use TLS 1.2 #>
    [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12

    <# account info #>
    $accessId = 'oiajweoriajweorijafake'
    $accessKey = 'L+C-9Sp2w2R)wA4Ufakefakefake'
    $company = 'fakecompany'

    $httpVerb = [string]$req.Method
    $resourcePath = $req.RequestUri.PathAndQuery

    <# Get current time in milliseconds #>
    $epoch = [Math]::Round((New-TimeSpan -start (Get-Date -Date "1/1/1970") -end (Get-Date).ToUniversalTime()).TotalMilliseconds)
    
    <# Concatenate Request Details #>
    $requestVars = $httpVerb + $epoch + $resourcePath
    write-verbose "Request Variables: $requestVars"

    <# Construct Signature #>
    $hmac = New-Object System.Security.Cryptography.HMACSHA256
    $hmac.Key = [Text.Encoding]::UTF8.GetBytes($accessKey)
    $signatureBytes = $hmac.ComputeHash([Text.Encoding]::UTF8.GetBytes($requestVars))
    $signatureHex = [System.BitConverter]::ToString($signatureBytes) -replace '-'
    $signature = [System.Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($signatureHex.ToLower()))

    <# Construct Headers #>
    $auth = 'LMv1 ' + $accessId + ':' + $signature + ':' + $epoch
    $req.headers.Authorization = $auth

    # call the next step in the Pipeline
    $responseTask = $next.SendAsync($req, $callback)
 
    # wait on the responseTask
    $response = $responseTask.Result
    
    # let the cmdlet finish it's job
    return $responseTask
 }
```

**Comment by rmustard** [comment]
Is anyone able to generate a powershell API from https://github.com/MicrosoftDocs/vsts-rest-api-specs?
I have tried removing any "x-ms-*" attributes, but still get the "anyOf" error.
```
ERROR: Schema violation: Data does not match any schemas from 'anyOf'
    - file:///C:/Users/rmustard/Documents/GitHub/vsts-rest-api-specs/specification/account/4.1/accounts.json ($)
```
Full output included here.

```
PS ~\GitHub\vsts-rest-api-specs\specification\account\4.1> autorest --powershell --input-file:accounts.json
AutoRest code generation utility [version: 3.0.5195; node: v10.15.1, max-memory: 8192 gb]
(C) 2018 Microsoft Corporation.
https://aka.ms/autorest
   Loading AutoRest core      'C:\Users\rmustard\.autorest\@microsoft.azure_autorest-core@3.0.5445\node_modules\@microsoft.azure\autorest-core\dist' (3.0.5445)
   Installing AutoRest extension '@microsoft.azure/autorest.powershell' (beta)
   Installed AutoRest extension '@microsoft.azure/autorest.powershell' (beta->2.0.423)
   Installing AutoRest extension '@microsoft.azure/autorest.remodeler' (beta)
   Installed AutoRest extension '@microsoft.azure/autorest.remodeler' (beta->2.0.254)
   Installing AutoRest extension '@microsoft.azure/autorest.csharp-v2' (beta)
   Installed AutoRest extension '@microsoft.azure/autorest.csharp-v2' (beta->2.0.276)
ERROR: Schema violation: Additional properties not allowed: x-ms-vss-area
    - file:///C:/Users/rmustard/Documents/GitHub/vsts-rest-api-specs/specification/account/4.1/accounts.json ($)
ERROR: Schema violation: Additional properties not allowed: x-ms-vss-resource-area-id
    - file:///C:/Users/rmustard/Documents/GitHub/vsts-rest-api-specs/specification/account/4.1/accounts.json ($)
ERROR: Schema violation: Data does not match any schemas from 'anyOf'
    - file:///C:/Users/rmustard/Documents/GitHub/vsts-rest-api-specs/specification/account/4.1/accounts.json ($)
ERROR: Schema violation: Additional properties not allowed: x-ms-docs-override-version
    - file:///C:/Users/rmustard/Documents/GitHub/vsts-rest-api-specs/specification/account/4.1/accounts.json ($)
ERROR: Schema violation: Additional properties not allowed: x-ms-vss-resource
    - file:///C:/Users/rmustard/Documents/GitHub/vsts-rest-api-specs/specification/account/4.1/accounts.json ($)
ERROR: Schema violation: Additional properties not allowed: x-ms-vss-method
    - file:///C:/Users/rmustard/Documents/GitHub/vsts-rest-api-specs/specification/account/4.1/accounts.json ($)
C:\Users\rmustard\.autorest\@microsoft.azure_autorest-core@3.0.5445\node_modules\@microsoft.azure\autorest-core\dist\lib\pipeline\pipeline.js - FAILURE {"exitCode":1}
Process() cancelled due to exception : [OperationAbortedException] Error occurred. Exiting. / Error: [OperationAbortedException] Error occurred. Exiting.
    at C:\Users\rmustard\.autorest\@microsoft.azure_autorest-core@3.0.5445\node_modules\@microsoft.azure\autorest-core\dist\lib\pipeline\schema-validation.js:30:19
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

**Comment by kevinblumenfeld** [comment]
Referencing issue: https://github.com/Azure/autorest.powershell/issues/385 
`ERROR (UnknownJsonType/LLCS2000): Schema 'BasicAuthApplicationName' is declared with invalid type 'undefined'`

**Comment by epanipinto-jc** [comment]
Our API requires an api key in the request header in order to authenticate. Is this supported? If so could you provide an example of how you would populate the api key?

```
"securityDefinitions": {
        "x-api-key": {
            "name": "x-api-key",
            "type": "apiKey",
            "in": "header"
        }
    },
```

**Comment by fearthecowboy** [comment]
@epanipinto-jc  -- we currently don't handle adding auth data automatically. An example how to add it manually https://github.com/Azure/autorest/tree/master/docs/powershell/samples/timeswire

**Comment by epanipinto-jc** [comment]
@fearthecowboy 

Not being too familiar with C# it took a bit to figure out but changing the namespace and updating this section to append to the header vs query string appears to have worked. Thanks for the tip and quick response!

`protected async Task<HttpResponseMessage> AddApiKey(HttpRequestMessage request, IEventListener callback, ISendAsync next)
{
    // add on the apikey we'll cheat and pull it from the environment :D
    request.Headers.Add("x-api-key",System.Environment.GetEnvironmentVariable("xApiKey"));
    // let it go on.
    return await next.SendAsync(request, callback);
}`

**Comment by epanipinto-jc** [comment]
It appears the generator only create cmdlets from items under the `paths` property within the OAS spec. Is there a way to create cmdlets for items that exist under `definitions` property that would essentially be `New-*` that would assist in the creation of objects? 

An example of what I would like to do is I have a path that is used for the creation of new users, lets call that cmdlet `Invoke-UserPost`. That endpoint takes in specific properties that define a user which is defined within the definitions section of the OAS spec. Ideally the generator would create a function called `New-User` based upon that definition that would return the user as an object so that I can pipe it to the first command. `New-User -FirstName:('Bruce') -LastName:('Wayne') | Invoke-UserPost`

**Comment by fearthecowboy** [comment]
We originally were going to make something like that, but it ended up being less-than-awesome.

The fallback at this point is that all the objects have typeconverters that allow you to pass in a `PSObject ` or `HashTable` that looks like the object 

ie, 

```  powershell

$user = @{ FirstName="Bruce"; LastName="Wayne" } | Invoke-UserPost 

```

**Comment by epanipinto-jc** [comment]
Ah bummer... I see how that could be complex. I was hoping to leverage the parameter validation and structure already in the OAS spec to remove redundant work. I stumbled on another way to kinda accomplish this, that is pretty manual, but i'm still trying to get it to work. I'm curious if you have any thoughts on this.

```
Import-Module ('TestApi.psd1') -Force
$NewUser = [TestApi.Models.UserPost]::new()
$NewUser.Firstname = 'Bruce'
$NewUser.Lastname = 'Wayne'
$NewUser.Email = 'NotBatman@WayneCorp.com'
$NewPhoneNumber = [TestApi.Models.UserPostPhoneNumbersItem]::new()
$NewPhoneNumber.Number = 'Work'
$NewPhoneNumber.Type = '123456789'
$NewUser.PhoneNumbers = $NewPhoneNumber
$NewUser | Invoke-UserPost
```

**Comment by fearthecowboy** [comment]
What's the difference between that and

```
Import-Module ('TestApi.psd1') -Force
$NewUser =  @{
  Firstname = 'Bruce'
  Lastname = 'Wayne'
  Email = 'NotBatman@WayneCorp.com'
  PhoneNumbers = @( @{
    Number = 'Work'
    Type = '123456789'
   })
}


$NewUser | Invoke-UserPost

# you can also cast it to force the typeconverter to run and give a typed obecjt
$TypedNewUser = [TestApi.Models.UserPost] $NewUser


```

Validation is currently only leveraged on `_expanded ` parameter variants (which explode out all the body properties into powershell parameters)

**Comment by fearthecowboy** [comment]
FWIW, You can also call the `Validate` method on an object an have it give the validation errors for it (which are derived from the original OAI doc)

We didn't end up using that in the model objects, but it should be usable. 

``` powershell
$NewUser = [TestApi.Models.UserPost] @{
  Firstname = 'Bruce'
  Lastname = 'Wayne'
  Email = 'NotBatman@WayneCorp.com'
  PhoneNumbers = @( @{
    Number = 'Work'
    Type = '123456789'
   })
}

$NewUser.Validate( $listener) 
```

the $listener needs to implement the `IEventListener` interface ... I don't have a sample for that, but you could spelunk the code and find it.

**Comment by epanipinto-jc** [comment]
@fearthecowboy, thanks for the help im still learning and figuring this out. Also maybe I am going about this the wrong way but in the end what I'm trying to achieve is to be able to transform our spec into cmdlets that enable our end users to quickly and easily write scripts to call our API in PowerShell. One aspect of this is to be able to create the objects that get sent to the cmdlets without having to worry about incorrectly building the objects.

After playing around some more this morning I am starting to realize maybe the class does not have all of the data that I thought it would. When I am talking about validation I want to leverage the bits in the spec that put requirements around things like `type`, `maxLength`, `enum` values, etc... What the difference is between our two examples, at least what I thought the difference was but I think I might be wrong, is that by creating the new object that the class would retain the requirements set in the spec and that when I populated the values of the new object that I would be prevented from adding invalid values or an invalid structure. I've included part of the specific definition from our spec as an example.

### Spec Definition

``` json
  "definitions": {
    "systemuserputpost": {
      "title": "SystemUserPost",
      "type": "object",
      "properties": {
        "username": {
          "type": "string"
        },
        "firstname": {
          "type": "string"
        },
        "lastname": {
          "type": "string"
        },
        "email": {
          "type": "string",
          "maxLength": 1024
        },
        "phoneNumbers": {
          "type": "array",
          "items": {
            "type": "object",
            "properties": {
              "type": {
                "type": "string",
                "enum": [
                  "home",
                  "work",
                  "mobile"
                ]
              },
              "number": {
                "type": "string",
                "maxLength": 10
              }
            }
          }
        }
      },
      "required": [
        "email",
        "username"
      ]
    }
  }
```

### Example 1

``` powershell
$NewPhoneNumber = [TestApi.Models.SystemuserputpostPhoneNumbersItem]::new()
$NewPhoneNumber.Type = 'home' # Should work because "home" exists in the enum array
$NewPhoneNumber.Type = 'NotHome' # Should not work because "NotHome" does not exists in the enum array, but it does which is not the desired result.
```

### Example 2

``` powershell
$NewPhoneNumber = [TestApi.Models.SystemuserputpostPhoneNumbersItem]::new()
$NewPhoneNumber.InvalidProperty = 'InvalidValue' # Will not work because "NewPhoneNumber" does not have a property called "InvalidProperty" which is the desired result.
```

VS

``` powershell
$NewUser = [TestApi.Models.SystemuserputpostPhoneNumbersItem]@{
    InvalidProperty = 'InvalidValue'  # Should not work because "NewPhoneNumber" does not have a property called "InvalidProperty" , but it does which is not the desired result.
    Username        = 'NotBatman'
    Firstname       = 'Bruce'
    Lastname        = 'Wayne'
    Email           = 'NotBatman@WayneCorp.com'
    PhoneNumbers    = @( @{
            Type   = 'home'
            Number = '123456789'
        })
}
```



Additionally is there a way to see the body or headers of the request being made or any info about the response? I only see the request url when I add the `-Debug` parameter.

**Comment by peombwa** [comment]
Is there a clear guide that shows how to use `profiles` in conjunction with `tag:all-api-versions` to generate a modules that targets two different API versions?

My overall goal is to generate single PowerShell module that can support two API versions. I'm generating my cmdlets using an OpenAPI v3 document, and my API has two versions, `beta` and `v1.0` - https://graph.microsoft.com/{API-VERSION}/.

**Comment by darrelmiller** [comment]
@fearthecowboy Do you have any docs/examples available for this?  We would like to use this for Microsoft Graph to switch between V1.0 and beta.

**Comment by fearthecowboy** [comment]
I don't have a doc for that at this point. 

API version is a PATH parameter? .... hmm.  I gotta go check how that would impact it.  

We should probably talk about how the different API versions work; it may be possible to craft something you can use as a profile; but it may require some tweaking to get it there. 

Can we meet on teams (sometime early tomorrow?).

**Comment by darrelmiller** [comment]
Sure, happy to meet and discuss.

**Comment by wsmelton** [comment]
I'm trying out autorest generator for PowerShell on [a swagger file](https://updates.thycotic.net/secretserver/restapiguide/WinAuth/swagger.json) for the company I work for...but getting an error that seems to point to the way the parameters are defined, but the way I read the specifications for OpenAPI it is correct:

The message I'm getting as errors all have the common:

> "Schema violation: Data does not match any schemas from 'oneOf'

![image](https://user-images.githubusercontent.com/11204251/74501761-ba98ad00-4eb0-11ea-9883-8338087bbc0d.png)

But if I look at the highlighted one in the image (see below) it all looks fine. Am I missing something obvious?

```
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "description": "User ID",
            "required": true,
            "type": "integer",
            "format": "int32"
          },
          {
            "name": "filter",
            "in": "query",
            "description": "Additional query options",
            "required": false
          },
          {
            "name": "skip",
            "in": "query",
            "description": "Number of records to skip before taking results",
            "required": false,
            "type": "integer",
            "format": "int32"
          },
          {
            "name": "take",
            "in": "query",
            "description": "Maximum number of records to include in results",
            "required": false,
            "type": "integer",
            "format": "int32"
          },
          {
            "name": "sortBy[n].name",
            "in": "query",
            "description": "Sort field name",
            "required": true,
            "type": "string"
          },
          {
            "name": "sortBy[n].direction",
            "in": "query",
            "description": "Sort direction",
            "required": true,
            "type": "string"
          },
          {
            "name": "sortBy[n].priority",
            "in": "query",
            "description": "Priority index. Sorts with lower values are executed earlier",
            "required": false,
            "type": "integer",
            "format": "int32"
          }
        ],
```

**Comment by epanipinto-jc** [comment]
Is there a way to implement a specified user agent with each API call? Would that be in the module.cs file?

**Comment by James-Allerton** [comment]
Hi,
I'm completely new to autorest and have tried it for the first time but when running the following command:
autorest --powershell --input-file=vecta.json
I get the Error:
FATAL: Error: Operation http-operation:0 : /api/Activities/{activityId} has no servers entry (Are you missing 'host' or 'servers' information in your source file?)
  Error: Plugin remodeler reported failure.

The Swagger JSON  in the vecta.json file is from: https://api.vecta.net/swagger/V0/swagger.json

I don't get this error when running
autorest --csharp --input-file=vecta.json
autorest --python --input-file=vecta.json

They both generate without errors, which make me think it's something specific to the powershell generator but i just wanted to check if anyone could see something i'm missing?

Thank you for any suggestions.

**Comment by epanipinto-jc** [comment]
I am trying to set this up in a CI scenario using Azure Pipelines to generate, test, and pack the module to then get published to the PowerShell Gallery. Do yall have any examples or recommended suggestions for this?

**Comment by Kurlee** [comment]
I need a quick sanity check for my attempt to generate a powershell module for the VMWare Workstation Rest API. 

[**Swagger file**](https://vdc-download.vmware.com/vmwb-repository/dcr-public/e07e4652-f0f8-47e9-a608-9685820cd8f1/9b8c620a-a191-4236-8555-c3d546735536/swagger.json) 

**Autorest command:**
`autorest --powershell --input-file:.\vmware.json --output-folder:.\OutDir --namespace:My.Example.Namespace`

The VMWare swagger file has a misspelling (can you believe it?), and besides that, everything is working for the most part. I have attempted to generate a CSharp client, but a fatal error occurs due to the 'produces' value. PowerShell module generation and build completes successfully, however, there is no way to pass a URI to the command. Notice, there is no `host` field in the swagger file, and I think, that may be the issue. The only flags in my generated cmdlets are the proxy, pipeline related.  

Every command fails with :
`Invalid URI: The format of the URI could not be determined`

Looking for a sanity check to confirm that I am not doing anything stupid.

**Comment by epanipinto-jc** [comment]
Is there a way to leverage the values within the `enum` property of the swagger spec to populate a parameters `ValidateSet` options?

**Comment by mockjv** [comment]
@James-Allerton I'm also seeing this with the swagger v2 JSON exposed by TeamCity and I've tried to modify the JSON and cannot figure out what it doesn't like.  Did you ever find a resolution to this?

**Comment by JustinGrote** [comment]
@fearthecowboy may be time to enable Github Discussions to replace this issue thread?
https://docs.github.com/en/discussions

**Comment by kjacobsen** [comment]
Hey folks,

Getting some schema violation errors, is there any method to ignore these?

```
kjacobsen@Orbiting-Dragon temp % autorest --input-file=https://api.stoplight.io/v1/versions/BG4PXY5ZvdxL5AmYX/export/oas.yaml --powershell
AutoRest code generation utility [cli version: 3.2.1; node: v12.22.1, max-memory: 2048 MB]
(C) 2018 Microsoft Corporation.
https://aka.ms/autorest
   Loading AutoRest core      '/Users/kjacobsen/.autorest/@autorest_core@3.4.3/node_modules/@autorest/core/dist' (3.4.3)
INFORMATION:
The default version of @autorest/powershell has been bumped from 2.1+ to 3.0+.
 > If you still want to use 2.1+ version, please specify it with --use:@autorest/powershell@2.1.{x}, e.g 2.1.401.

INFORMATION: > Loading AutoRest extension '@autorest/powershell' (~3.0.0->3.0.430)
INFORMATION: > Loading AutoRest extension '@autorest/modelerfour' (4.15.414->4.15.414)
Process() cancelled due to failure
ERROR: Schema violation: should be array (paths > /companies/info > get > responses > 200 > schema > properties > authentications > items > properties > init_url > required)
    - https://api.stoplight.io/v1/versions/BG4PXY5ZvdxL5AmYX/export/oas.yaml:317:12 ($)
ERROR: Schema violation: should be array (paths > /companies/info > get > responses > 200 > schema > properties > authentications > items)
    - https://api.stoplight.io/v1/versions/BG4PXY5ZvdxL5AmYX/export/oas.yaml:304:9 ($)
ERROR: Schema violation: should match some schema in anyOf (paths > /companies/info > get > responses > 200 > schema > properties > authentications > items)
    - https://api.stoplight.io/v1/versions/BG4PXY5ZvdxL5AmYX/export/oas.yaml:304:9 ($)
ERROR: Schema violation: should NOT have additional properties (paths > /companies/info > get > responses > 200 > schema)
  additionalProperty: properties
    - https://api.stoplight.io/v1/versions/BG4PXY5ZvdxL5AmYX/export/oas.yaml:283:6 ($)
ERROR: Schema violation: should match exactly one schema in oneOf (paths > /companies/info > get > responses > 200 > schema)
  passingSchemas: null
    - https://api.stoplight.io/v1/versions/BG4PXY5ZvdxL5AmYX/export/oas.yaml:283:6 ($)
ERROR: Schema violation: should be equal to one of the allowed values (paths > /companies/info > get > responses > 200 > schema > type)
  allowedValues: file
    - https://api.stoplight.io/v1/versions/BG4PXY5ZvdxL5AmYX/export/oas.yaml:284:7 ($)
ERROR: Schema violation: should be array (paths > /companies/info > get > responses > 200 > schema > properties > authentications > items > properties > init_url > required)
    - https://api.stoplight.io/v1/versions/BG4PXY5ZvdxL5AmYX/export/oas.yaml:317:12 ($)
ERROR: Schema violation: should be array (paths > /companies/info > get > responses > 200 > schema > properties > authentications > items)
    - https://api.stoplight.io/v1/versions/BG4PXY5ZvdxL5AmYX/export/oas.yaml:304:9 ($)
ERROR: Schema violation: should match some schema in anyOf (paths > /companies/info > get > responses > 200 > schema > properties > authentications > items)
    - https://api.stoplight.io/v1/versions/BG4PXY5ZvdxL5AmYX/export/oas.yaml:304:9 ($)
ERROR: Schema violation: should NOT have additional properties (paths > /companies/info > get > responses > 200 > schema)
  additionalProperty: properties
    - https://api.stoplight.io/v1/versions/BG4PXY5ZvdxL5AmYX/export/oas.yaml:283:6 ($)
ERROR: Schema violation: should match exactly one schema in oneOf (paths > /companies/info > get > responses > 200 > schema)
  passingSchemas: null
    - https://api.stoplight.io/v1/versions/BG4PXY5ZvdxL5AmYX/export/oas.yaml:283:6 ($)
ERROR: Schema violation: should be equal to one of the allowed values (paths > /companies/info > get > responses > 200 > schema > type)
  allowedValues: file
    - https://api.stoplight.io/v1/versions/BG4PXY5ZvdxL5AmYX/export/oas.yaml:284:7 ($)
```

**Comment by VencislavKolev** [comment]
Hi guys,
I'm using autorest to generate powershell cmdlets from two (2) swagger definitions which share the same host name, but different base paths and _some of the endpoints in both API specs are identitical_
example: 
host: `mypage.com`
file1 base path: `my/api`
file2 base path: `my/saas/api`
endpoint from file1 and file2: `users/{userId}/tasks/{taskId}`
endpoint from file1 and file2: `users/{userId}/tasks/`

I'm facing the following issues - generated c# cmdlets have the following names 
with endpoint for file 1 -` Get-Task_Get / Get-Task_List `
with endpoint for file 2 - `Get-Task_Get1 / Get-Task_List1`
I had to go through the generated code and write down which .cs file correlates to which base path, however this isn't my biggest concern.
The generated script has name of Get-Task.ps1, below are some of the possible input parameters. Now if I want to execute lets say `users/{userId}/tasks/` for base path `my/saas/api` I have comment out Get, Get1 and List1 from first parameter so it calls the desired endpoint. Whole point is I have to play around to comment and uncomment parameters based on what I want.
I would like to ask if there is a configuration I can write in my readme.md config file before running the **autorest** command or the only solution is to manually create separate **ps1** with whats neccessary and give it a correct naming?

Thanks
  
 >Code
        # [Parameter(ParameterSetName='Get', Mandatory)]
        # [Parameter(ParameterSetName='Get1', Mandatory)]
        [Parameter(ParameterSetName = 'List', Mandatory)]
        #  [Parameter(ParameterSetName='List1', Mandatory)]
        [Sample.API.Category('Path')]
        [System.String]
        ${userId},
        # [Parameter(ParameterSetName = 'Get', Mandatory)]
        # [Parameter(ParameterSetName = 'Get1', Mandatory)]
         [Sample.API.Category('Path')]
         [System.String]
         ${taskId},

**Comment by evochriso** [comment]
Hi,
When running the build-module.ps1 script, I get several errors:
- CS1073: Unexpected token '$' 
- CS1525: Invalid expression term ''

Checking the lines its complaining about, they all contain this:
var _url = new global::System.Uri($"{$host}{pathAndQuery}");

Which makes sense, **$**host is invalid C# syntax. Why would the generator write '$host' instead of just 'host'?

Thanks!



Type       Extension Name                           Version      Location
 core       @autorest/core                           3.0.6375     C:\Users\Chris\.autorest\@autorest_core@3.0.6375
 core       @autorest/core                           3.8.4        C:\Users\Chris\.autorest\@autorest_core@3.8.4
 extension  @autorest/modelerfour                    4.15.414     C:\Users\Chris\.autorest\@autorest_modelerfour@4.15.414
 extension  @autorest/powershell                     3.0.481      C:\Users\Chris\.autorest\@autorest_powershell@3.0.481

**Comment by JnaneshD05** [comment]
Is there any steps to publish the generated modules to gallery? Please share

**Comment by dicolanl** [comment]
Hi
Does anyone have a sample for handling response?
eg we want to change some fields that are returned from a int enum to a string or change a timestamp from epoch to something human readable.

**Comment by arnydo** [comment]
Does anyone have any tips on addressing this error:

```
error   | Modeler | Boolean schema 'schemas:657' with unknown format: 'bool' is not valid
error   | Modeler/UnknownSchemaType | The model schemas:657 does not have a recognized schema type 'boolean'
```

**Comment by IzaacJ** [comment]
New to autorest, and am running into issues regarding circular references, and am completely lost on how I setup the directive no-inline thing. Is it in the swagger.json, if so, where, or is it in another file, and if so, where and what name should it have?
Been searching through issues both here and swagger, but unable to find any clear documentation for it.

**Comment by SamLW1** [comment]
> ~New to autorest, and am running into issues regarding circular references, and am completely lost on how I setup the directive no-inline thing. Is it in the swagger.json, if so, where, or is it in another file, and if so, where and what name should it have? Been searching through issues both here and swagger, but unable to find any clear documentation for it.~
> 
> Stumbled upon it when just skimming through the docs, for the nth time. https://github.com/Azure/autorest/blob/main/docs/generate/readme.md#keeping-your-options-in-one-place-the-preferred-option
> 
> Another issue has showed up thou, generated variable name starts with a number, and is therefore invalid C#. It is in `generated/cmdlets` directory, and it seems to have only happened in the only file that ends with `ViaIdentity` and a double-digit number. Any other generated cmdlets that have a double-digit number but without `ViaIdentity` does not suffer from this issue. None of the generated files with `ViaIdentity` and a single-digit number have this issue.
> 
> The generated code is below:
> 
> /// <summary>Identity Parameter</summary>
>         [global::System.Management.Automation.Parameter(Mandatory = true, HelpMessage = "Identity Parameter", ValueFromPipeline = true)]
>         [global::OurModule.Category(global::OurModule.ParameterCategory.Path)]
>         public OurModule.Models.IOurModuleIdentity 0InputObject { get => this._0InputObject; set => this._0InputObject = value; }
> Edit: Also noticed now that the `onCreated` and `onOk` parameters is duplicated in almost every method in `generated/api/OurModule.cs`. Probably due to a lot of the `generated/cmdlets/` files contain duplicate `overrideOnCreated`, `onCreated`, `overrideOnOk` and `onOk` with the same parameter types.
> 
> Edit2: Think I've figured out why there are duplicates of `onCreated` and `onOk`. Our API, build with Entity Framework, returns an object with error details (I'm unfamiliar with EF, this might be OOB or not) and in the swagger it correctly specifies that it is returned on a bunch of non 20* codes, but autorest generates a handler for those error objects for both 200 and 201, even if that isn't specified in the swagger.
> 
> Example endpoint that generates duplicate `onCreated` and `onOk` methods:
> 
>     "/api/Function": {
>       "get": {
>         "tags": [
>           "Function"
>         ],
>         "summary": "Gets all functions asynchronous.",
>         "parameters": [
>           {
>             "name": "ShowRemoved",
>             "in": "query",
>             "schema": {
>               "type": "boolean",
>               "default": false
>             }
>           }
>         ],
>         "responses": {
>           "200": {
>             "description": "Success",
>             "content": {
>               "application/json": {
>                 "schema": {
>                   "type": "array",
>                   "items": {
>                     "$ref": "#/components/schemas/FunctionViewListModel"
>                   }
>                 }
>               }
>             }
>           },
>           "204": {
>             "description": "No Content"
>           },
>           "400": {
>             "description": "Bad Request",
>             "content": {
>               "application/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               }
>             }
>           },
>           "401": {
>             "description": "Unauthorized",
>             "content": {
>               "application/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               }
>             }
>           },
>           "403": {
>             "description": "Forbidden",
>             "content": {
>               "application/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               }
>             }
>           },
>           "404": {
>             "description": "Not Found",
>             "content": {
>               "application/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               }
>             }
>           },
>           "422": {
>             "description": "Client Error",
>             "content": {
>               "application/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               }
>             }
>           }
>         }
>       },
>       "post": {
>         "tags": [
>           "Function"
>         ],
>         "summary": "Creates a function asynchronous.",
>         "requestBody": {
>           "description": "The create function model.",
>           "content": {
>             "application/json": {
>               "schema": {
>                 "$ref": "#/components/schemas/FunctionCreateModel"
>               }
>             }
>           }
>         },
>         "responses": {
>           "201": {
>             "description": "Created",
>             "content": {
>               "text/plain": {
>                 "schema": {
>                   "type": "string"
>                 }
>               },
>               "application/json": {
>                 "schema": {
>                   "type": "string"
>                 }
>               },
>               "text/json": {
>                 "schema": {
>                   "type": "string"
>                 }
>               }
>             }
>           },
>           "400": {
>             "description": "Bad Request",
>             "content": {
>               "text/plain": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               },
>               "application/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               },
>               "text/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               }
>             }
>           },
>           "401": {
>             "description": "Unauthorized",
>             "content": {
>               "text/plain": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               },
>               "application/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               },
>               "text/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               }
>             }
>           },
>           "403": {
>             "description": "Forbidden",
>             "content": {
>               "text/plain": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               },
>               "application/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               },
>               "text/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               }
>             }
>           },
>           "404": {
>             "description": "Not Found",
>             "content": {
>               "text/plain": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               },
>               "application/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               },
>               "text/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               }
>             }
>           },
>           "422": {
>             "description": "Client Error",
>             "content": {
>               "text/plain": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               },
>               "application/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               },
>               "text/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/ProblemDetails"
>                 }
>               }
>             }
>           }
>         }
>       }
>     },

Did you get any further with the duplicate 'onOK' etc issue? We are experiencing the exact same problem.

**Comment by JustinGrote** [comment]
> @epanipinto-jc -- we currently don't handle adding auth data automatically. An example how to add it manually [Azure/autorest@`master`/docs/powershell/samples/timeswire](https://github.com/Azure/autorest/tree/master/docs/powershell/samples/timeswire)

These examples are completely gone from the history, I think the repo was force pushed from master to main when renamed and history was lost.

Is there an example now that generates this correctly?

**Comment by IzaacJ** [comment]
@SamLW1 Sorry for a late reply, but we abandoned using this and switched to using openapi-generator instead, and heavily customized the templates. Was quite a bit of work, but it has been running smoothly, building a working powershell module almost on a daily basis as a part of our cicd workflow.
